### PR TITLE
add a ddp friendly script

### PR DESCRIPTION
**Pull request by not-lain**
this pr introduces a new file to the repo that supports running on multiple GPUs by default levaraging the trainer api as the main key componenet.
I also made some minor enhancements in the new file in comparison to the main training script, the changes include : 
* remove unecessary extra pad token between input and output
* better pad token id replacement in label creation (using the attention mask to patch fast tokenizer not allowing for negative token ids) 
* minor model update with backward compatibility to support trainer api syntax 
* doubling the data collator as a data processor as well
* minor patch to the safetonsors save (i think this has been fixed in other prs) 
* padding only to match the longest sequence length instead of the max length (less resources if the batch has smaller length) 

running on kaggle because cat and multiple gpus 

image

### CODE REVIEW

**Comment by andimarafioti** [comment]
Wow, good work @not-lain , this is super cool! I think for the spirit of this repo though, the "pure pytorch" part is important. We use transformers for a few things (loading models from the hub and such), but it would be great if the ddp training was still pure torch. WDYT?
